Release redo

cevich opened the pull request:
The initial implementation was far more complicated than necessary.
Strip out the complexities in favor of a simpler and more direct
approach.

***Note***: The new cross-compile tasks will likely fail when this
merges, due to the container image not being built yet
(they're defined in this PR).  Re-running them after 30m or so
should allow them to pass.

cevich commented:
@jwhonce @baude @mheon the 'release' task is badly broken and far too complex.  This is the beginnings (somewhat incomplete / untested) re-implementation.  Key elements are:

* No more 'release' task
* Fixed & re-purposed (broken) 'brew-pkg' Makefile target for windows and darwin
* Produce the release archives directly in the testing tasks (cross-compile-test for windows|darwin)
* Use the just built & tested podman, with a purpose-built container, to upload the archives.
* For podman on linux, make a tarball
* For podman-remote on windows/darwin, use a zip file

PTAL at the Makefile - it seems to work for me.  Better this way?

Should I put back the old 'brew-pkg' target, and if so, does the archive-subdirectory have to be 'brew'?

jwhonce commented:
@baude , If @cevich is going through all this work should a linux podman-remote tarball also be created for completeness?

cevich commented:
> going through all this work

Arguably it's necessary work, the current implementation is too complicated and broken.  Simplification was demanded, but you're correct to ask "while I'm at it"...

> should a linux podman-remote tarball also be created for completeness?

At this stage, this would be a nearly trivial addition.

cevich commented:
@jwhonce PTAL, I got this working for PR's (and presumably) the master branch.  It uses tarballs for the libpod builds, and zip files for the remote client.

N/B: I re-purposed the `brew-pkg` for `podman-remote-$(RELEASE_NUMBER)-%.zip` target, is that something you (or anyone) might care about?  If so I can put it back, no prob.

rhatdan commented:
LGTM
@jwhonce @cevich is this ready to merge?

cevich commented:
/hold

cevich commented:
note to me:  spotted some bugs:
* filename extensions in storage are wrong (all are .zip)
* remote-client cross-target name lost in uploaded filename (fedora -> darwin)

jwhonce commented on contrib/upldrel/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+
+RUN yum -y update && \
+    yum -y install unzip && \
+    yum clean all
```
You should add an rpm -V unzip after the yum install.  Yum doesn't error out in all cases when an install fails.

cevich replied:
oh right, it's a good idea, I'll add that.

cevich commented:
/hold cancel

cevich commented:
okay, pretty sure I fixed both file-naming problems.  Also added Jhon's suggestion.  I wiped out the pr-release bucket, rebased and force-pushed.  Should be good to go now (assuming tests pass)

cevich commented on contrib/cirrus/build_release.sh:
```diff
@@ -0,0 +1,30 @@
+        windows) ;&
+        darwin)
+            make podman-remote-${CROSS_PLATFORM}-release
+            ;;
```
@jwhonce @ashley-cui I think the windows/darwin sections above are probably a good place to call into whatever is needed to produce msi and flatpack out of the release zip files `make` is going to emit.

Conveniently at this point in automation, a fully functional local podman is guaranteed.  So if your tooling can live in a container-image, it could easily be used here to do the business.  Come bother me when you're ready and I can assist with the integration.

jwhonce replied:
@cevich I'm using

libmsi1-0.99-2.fc30.x86_64
msitools-0.99-2.fc30.x86_64

I don't know if those are in your packaging container yet.  But the next inclusion if you could add those.

ashley-cui commented:
> @jwhonce PTAL, I got this working for PR's (and presumably) the master branch. It uses tarballs for the libpod builds, and zip files for the remote client.
> 
> N/B: I re-purposed the `brew-pkg` for `podman-remote-$(RELEASE_NUMBER)-%.zip` target, is that something you (or anyone) might care about? If so I can put it back, no prob.

Hi, just took a look at this and wanted to give a little context to the `brew-pkg` target. As long as the the zipped file that is released can be accessed/downloaded via a link, it should be okay, since homebrew needs to pull down this zip via an url. Currently I've just been manually uploading it as an asset for each release, but if this can be uploaded somewhere automatically, (not really familiar with releases at all) then it would be perfect! The archive-subdirectory does not have to be `brew`, but if that's changed let me know, so I can change the script in homebrew so that doesn't break. I'm also not to familiar with Makefiles, so if you can let me know the difference between how the re-written one works and the broken brew-pkg works, that would be helpful to make sure we have everything that needs to go into the brew pkg correctly. Thanks!

ashley-cui commented:
@cevich I would prefer if the subdir would not be named podman-remote, since from a user on the Mac's point of view, podman is just podman, and not very remote. Shouldn't be a big deal though, the version'd folder could work fine. 
I'm not really familiar with webhooks, but it would sound ideal if podman in homebrew could be updated every time a version is released. The webhook would need to edit a ruby file and open a PR for homebrew. If this is possible, let me know and I can take a look at it or something. (probably would need help though)

cevich commented:
> I would prefer if the subdir would not be named podman-remote, since from a user on the Mac's point of view, podman is just podman, and not very remote. Shouldn't be a big deal though, the version'd folder could work fine.

oh okay, no problem, I put it back to 'podman-$VERSION'.  I'd like to include the version number in the sub-dir name, since for humans that prevents clashes when unzipping.  However, it can be a pain for automation to deal with, since the number will inevitably change.

If you have a mac, take a pek and make sure this works: https://storage.cloud.google.com/libpod-pr-releases/libpod-remote-pr3710-darwin---amd64.zip

> I'm not really familiar with webhooks, but it would sound ideal if podman in homebrew could be 

They're simple in concept: It's just a URL that automation would send some custom JSON to.  In other words, it's just a REST API endpoint.   I'm not familiar with homebrew to know if this is supported or not.  It's fairly common for SaaS tools to have them, so worth checking their docs.

> updated every time a version is released. The webhook would need to edit a ruby file and open a PR for homebrew. If this is possible, let me know and I can take a look at it or something. (probably would need help though)

A webhook is a one-shot deal, though the JSON can be whatever is required.  If homebrew does not support webhooks, it *is* possible ti bolt-on automation to do what you describe.  The down-sides are:

* It doesn't care if there's a human on the other end or not.  It will gladly open up 10,000 PRs.  This could be limited to only running for tagged-releases (every few weeks), but it still will not care if it's 1 or 100 (if something goes wrong).

* it's somewhat involved (from a security standpoint) to (safely) automate opening of PRs.  I know how to do it, but it's likely a week or two worth of work (including testing), so we should try to avoid this if possible.

baude commented:
/lgtm
